L1Trigger/TrackerDTC/plugins/ProducerDTC.cc

```diff
@@ -0,0 +1,114 @@
+#include "FWCore/Framework/interface/stream/EDProducer.h"
+#include "FWCore/Framework/interface/Run.h"
+#include "FWCore/Framework/interface/EventSetup.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/Utilities/interface/EDGetToken.h"
+#include "FWCore/Utilities/interface/EDPutToken.h"
+#include "FWCore/Utilities/interface/ESGetToken.h"
+#include "DataFormats/Common/interface/Handle.h"
+#include "DataFormats/DetId/interface/DetId.h"
+
+#include "DataFormats/L1TrackTrigger/interface/TTTypes.h"
+#include "DataFormats/L1TrackTrigger/interface/TTDTC.h"
+#include "L1Trigger/TrackTrigger/interface/Setup.h"
+#include "L1Trigger/TrackTrigger/interface/SensorModule.h"
+#include "L1Trigger/TrackerDTC/interface/LayerEncoding.h"
+#include "L1Trigger/TrackerDTC/interface/DTC.h"
+#include "L1Trigger/TrackerTFP/interface/DataFormats.h"
+
+#include <numeric>
+#include <algorithm>
+#include <vector>
+#include <string>
+#include <memory>
+
+using namespace std;
+using namespace edm;
+using namespace tt;
+using namespace trackerTFP;
+
+namespace trackerDTC {
+
+  /*! \class  trackerDTC::ProducerDTC
+   *  \brief  Class to produce hardware like structured TTStub Collection used by Track Trigger emulators
+   *  \author Thomas Schuh
+   *  \date   2020, Jan
+   */
+  class ProducerDTC : public stream::EDProducer<> {
+  public:
+    explicit ProducerDTC(const ParameterSet&);
+    ~ProducerDTC() override {}
+
+  private:
+    void produce(Event&, const EventSetup&) override;
+    // ED input token of TTStubs
+    EDGetTokenT<TTStubDetSetVec> edGetToken_;
+    // ED output token for accepted stubs
+    EDPutTokenT<TTDTC> edPutTokenAccepted_;
+    // ED output token for lost stubs
+    EDPutTokenT<TTDTC> edPutTokenLost_;
+    // Setup token
+    ESGetToken<Setup, SetupRcd> esGetTokenSetup_;
+    // DataFormats token
+    ESGetToken<DataFormats, DataFormatsRcd> esGetTokenDataFormats_;
+    // LayerEncoding token
+    ESGetToken<LayerEncoding, SetupRcd> esGetTokenLayerEncoding_;
+  };
+
+  ProducerDTC::ProducerDTC(const ParameterSet& iConfig) {
+    // book in- and output ED products
+    const auto& inputTag = iConfig.getParameter<InputTag>("InputTag");
+    const auto& branchAccepted = iConfig.getParameter<string>("BranchAccepted");
+    const auto& branchLost = iConfig.getParameter<string>("BranchLost");
+    edGetToken_ = consumes<TTStubDetSetVec>(inputTag);
+    edPutTokenAccepted_ = produces<TTDTC>(branchAccepted);
+    edPutTokenLost_ = produces<TTDTC>(branchLost);
+    // book ES products
+    esGetTokenSetup_ = esConsumes<Setup, SetupRcd, Transition::BeginRun>();
+    esGetTokenDataFormats_ = esConsumes<DataFormats, DataFormatsRcd, Transition::BeginRun>();
+    esGetTokenLayerEncoding_ = esConsumes<LayerEncoding, SetupRcd, Transition::BeginRun>();
```

Comment:
These should now be
```suggestion
    esGetTokenSetup_ = esConsumes<Setup, SetupRcd>();
    esGetTokenDataFormats_ = esConsumes<DataFormats, DataFormatsRcd>();
    esGetTokenLayerEncoding_ = esConsumes<LayerEncoding, SetupRcd>();
```
because the data products are obtained in the `produce()` function.

They could also be simplified further
```suggestion
    esGetTokenSetup_ = esConsumes();
    esGetTokenDataFormats_ = esConsumes();
    esGetTokenLayerEncoding_ = esConsumes();
```
(similarly the `consumes()` and `produces()` calls above can omit the types; the types are deduced from the tokens)

(from an error pointed out to me privately by @BenjaminRS)

Reply:
done